### PR TITLE
[Improvement](aggregation) optimization for aggregation hash_table_lazy_emplace

### PR DESCRIPTION
**Pull request by BiteTheDDDDt**
## Proposed changes
```
select count(1) from (
select ss_customer_sk customer_sk
      ,ss_item_sk item_sk
from store_sales,date_dim
where ss_sold_date_sk = d_date_sk
  and d_month_seq between 1199 and 1199 + 11 and ss_sold_date_sk IS NOT NULL
group by ss_customer_sk
        ,ss_item_sk
) a;

pipeline=false,nereids=false(agg)
                        11s386ms
                              -  GetResultsTime:  713.536ms
                              -  HashTableComputeTime:  10s163ms
                              -  HashTableInputCount:  55.285583M  (55285583)
                              -  HashTableIterateTime:  279.130ms
                              -  HashTableSize:  54.116764M  (54116764)
                              -  InsertKeysToColumnTime:  401.314ms

                        8s761ms
                              -  GetResultsTime:  422.797ms
                              -  HashTableComputeTime:  7s898ms
                              -  HashTableEmplaceTime:  7s485ms
                              -  HashTableInputCount:  55.285586M  (55285586)
                              -  HashTableIterateTime:  270.700ms
                              -  HashTableSize:  54.116764M  (54116764)
                              -  InsertKeysToColumnTime:  112.495ms

pipeline=true,nereids=true(distinct agg + two phase agg + shared scan)
                        12s872ms
                          -  HashTableComputeTime:  9s738ms
                          -  HashTableInputCount:  55.286154M  (55286154)
                          -  HashTableIterateTime:  0ns
                          -  HashTableSize:  54.116764M  (54116764)

                        10s789ms
                          -  HashTableComputeTime:  7s365ms
                          -  HashTableEmplaceTime:  7s19ms
                          -  HashTableInputCount:  55.286154M  (55286154)
                          -  HashTableIterateTime:  0ns
                          -  HashTableSize:  54.116764M  (54116764)

```
## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run p0

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.01 seconds
 stream load tsv:          505 seconds loaded 74807831229 Bytes, about 141 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.5 seconds inserted 10000000 Rows, about 338K ops/s
 storage size: 17170491643 Bytes

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 47.06 seconds
 stream load tsv:          508 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.4 seconds inserted 10000000 Rows, about 340K ops/s
 storage size: 17167764621 Bytes

File: be/src/vec/common/aggregation_common.h [hunk 1]
```diff
@@ -218,6 +196,112 @@ T pack_fixed(size_t i, size_t keys_size, const ColumnRawPtrs& key_columns, const
...
+    T zero {};
+    for (size_t j = 0; j < key_columns.size(); ++j) {
+        const char* data = key_columns[j]->get_raw_data().data;
+
+        if (key_sizes[j] == 1) {
+            if (nullmap_columns.size() && nullmap_columns[j]) {
+                const auto& nullmap =
+                        assert_cast<const ColumnUInt8&>(*nullmap_columns[j]).get_data().data();
+                for (size_t i = 0; i < row_numbers; ++i) {
+                    memcpy_fixed_1((char*)(&result[i]) + offset,
```
Comment (HappenLee) [inline be/src/vec/common/aggregation_common.h:235]: why not use template to replace the 1,2,4,8,16？

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run ckbench

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.64 seconds
 stream load tsv:          526 seconds loaded 74807831229 Bytes, about 135 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.7 seconds inserted 10000000 Rows, about 336K ops/s
 storage size: 17161911528 Bytes

File: be/src/vec/common/aggregation_common.h [hunk 2]
```diff
@@ -218,6 +196,70 @@ T pack_fixed(size_t i, size_t keys_size, const ColumnRawPtrs& key_columns, const
...
+        const char* data = key_columns[j]->get_raw_data().data;
+
+        auto foo = [&]<typename Fixed>(Fixed zero) {
+            if (nullmap_columns.size() && nullmap_columns[j]) {
+                const auto& nullmap =
+                        assert_cast<const ColumnUInt8&>(*nullmap_columns[j]).get_data().data();
+                for (size_t i = 0; i < row_numbers; ++i) {
+                    // make sure null cell is filled by 0x0
+                    memcpy_fixed<Fixed>((char*)(&result[i]) + offset,
+                                        nullmap[i] ? (char*)&zero : data + i * key_sizes[j]);
```
Comment (HappenLee) [inline be/src/vec/common/aggregation_common.h:235]: `key_sizes[j]` replace by `sizeof(Fixed)`

**Comment by BiteTheDDDDt** [comment]
run p0

**Review by HappenLee — DISMISSED** [review]
LGTM

File: be/src/vec/common/columns_hashing.h [hunk 3]
```diff
@@ -73,6 +75,14 @@ struct HashMethodOneNumber : public columns_hashing_impl::HashMethodBase<
     FieldType get_key_holder(size_t row, Arena&) const {
         return unaligned_load<FieldType>(vec + row * sizeof(FieldType));
     }
+
+    std::vector<FieldType> get_keys(size_t rows_number) const {
```
Comment (HappenLee) [inline be/src/vec/common/columns_hashing.h:79]: std::span

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 47.02 seconds
 stream load tsv:          512 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          29.2 seconds inserted 10000000 Rows, about 342K ops/s
 storage size: 17162063165 Bytes

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 47.17 seconds
 stream load tsv:          509 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          28.9 seconds inserted 10000000 Rows, about 346K ops/s
 storage size: 17162578239 Bytes

**Comment by BiteTheDDDDt** [comment]
run buildall

**Review by HappenLee — APPROVED** [review]
LGTM

File: be/src/vec/exec/distinct_vaggregation_node.cpp [hunk 4]
```diff
@@ -53,23 +53,23 @@ Status DistinctAggregationNode::_distinct_pre_agg_with_serialized_key(
...
 
     int rows = in_block->rows();
-    IColumn::Selector distinct_row;
-    distinct_row.reserve(rows);
+    _distinct_row.clear();
+    _distinct_row.reserve(rows);
 
     RETURN_IF_CATCH_EXCEPTION(
-            _emplace_into_hash_table_to_distinct(distinct_row, key_columns, rows));
+            _emplace_into_hash_table_to_distinct(_distinct_row, key_columns, rows));
```
Comment (zhangstar333) [inline be/src/vec/exec/distinct_vaggregation_node.cpp:60]: maybe could add those optimization on partition_sort_node hash map ,
`void VPartitionSortNode::_emplace_into_hash_table(const ColumnRawPtrs& key_columns,`

**Review by zhangstar333 — APPROVED** [review]

**Pull request merged**